File: acceleration.py
"""
Volume Acceleration Handler.

Provides acceleration-based volume control that increases the volume
change rate based on how fast the user is rotating the knob.
"""

from typing import List


class AccelerationHandler:
    """
    Handles volume acceleration based on click speed.

    When the user rotates the knob quickly, the volume changes faster.
    The acceleration curve is configurable via volume_increases_list.
    """

    def __init__(self, min_click: float, max_per_click_avg: float, volume_list: List[int]):
        """
        Initialize the acceleration handler.

        Args:
            min_click: Minimum time between clicks to consider them separate (seconds)
            max_per_click_avg: Maximum average time per click for acceleration (seconds)
            volume_list: List of volume increments for each acceleration level
        """
        self.min_click = min_click
        self.max_per_click_avg = max_per_click_avg
        self.volume_increases_list = volume_list
        self.len = len(volume_list)  # Cache length (list is immutable)
        self.last_button = 0
        self.last_time = 0
        self.first_time = 0
        self.distance = 0
        self.count = 1
        self.delta_time = 0

    def calculate_speed(self, current_time: float, button: int) -> int:
        """
        Calculate volume change based on click speed.

        Args:
            current_time: Current timestamp
            button: Button/direction identifier (to detect direction changes)

        Returns:
            Volume delta (how much to change volume by)
        """
        self.delta_time = current_time - self.last_time
        # Guard against division by zero (shouldn't happen with count initialized to 1)
        if self.count > 0:
            avg_step_time = (current_time - self.first_time) / self.count
        else:
            avg_step_time = float('inf')

        if (self.last_button != button) or (avg_step_time > self.max_per_click_avg) or (self.delta_time > self.min_click):
            self.distance = 1
            self.count = 1
            self.first_time = current_time
        else:
            if self.count <= self.len:  # count 1..len maps to indices 0..len-1
                self.distance = self.volume_increases_list[self.count - 1]
            else:
                self.distance = self.volume_increases_list[-1]
            self.count += 1
        self.last_button = button
        self.last_time = current_time
        return int(self.distance)

File: go/hid/acceleration.go
package hid

// AccelerationHandler provides speed-dependent volume acceleration.
// When the user rotates the knob quickly, volume changes faster.
type AccelerationHandler struct {
	minClick       float64 // Min time between clicks to consider separate (seconds)
	maxPerClickAvg float64 // Max average time per click for acceleration
	volumeList     []int   // Volume increments per acceleration level
	listLen        int

	lastButton int
	lastTime   float64
	firstTime  float64
	count      int
}

// NewAccelerationHandler creates a handler with the given acceleration curve.
func NewAccelerationHandler(minClick, maxPerClickAvg float64, volumeList []int) *AccelerationHandler {
	return &AccelerationHandler{
		minClick:       minClick,
		maxPerClickAvg: maxPerClickAvg,
		volumeList:     volumeList,
		listLen:        len(volumeList),
		count:          1,
	}
}

// CalculateSpeed returns the volume delta based on rotation speed.
// button identifies the direction (to detect direction changes).
func (a *AccelerationHandler) CalculateSpeed(currentTime float64, button int) int {
	deltaTime := currentTime - a.lastTime

	var avgStepTime float64
	if a.count > 0 {
		avgStepTime = (currentTime - a.firstTime) / float64(a.count)
	} else {
		avgStepTime = 1e9 // effectively infinity
	}

	var distance int

	if a.lastButton != button || avgStepTime > a.maxPerClickAvg || deltaTime > a.minClick {
		// Reset: direction change, too slow on average, or long gap
		distance = 1
		a.count = 1
		a.firstTime = currentTime
	} else {
		// Accelerate: 1st click always resets to 1, then list[count] from 2nd click.
		// list[0] = 2nd click delta, list[1] = 3rd click delta, etc.
		// Beyond list length, last element repeats (caps acceleration).
		idx := a.count
		if idx >= a.listLen {
			idx = a.listLen - 1
		}
		distance = a.volumeList[idx]
		a.count++
	}

	a.lastButton = button
	a.lastTime = currentTime
	return distance
}
